applications/CompressiblePotentialFlowApplication/custom_elements/compressible_perturbation_potential_flow_element.cpp

```diff
@@ -553,6 +542,47 @@ void CompressiblePerturbationPotentialFlowElement<Dim, NumNodes>::CalculateRight
     }
 }
 
+template <int Dim, int NumNodes>
+void CompressiblePerturbationPotentialFlowElement<Dim, NumNodes>::CalculateLeftHandSideContribution(
+    BoundedMatrix<double, NumNodes, NumNodes>& rLhs_total,
+    const ProcessInfo& rCurrentProcessInfo,
+    const array_1d<double, Dim>& rVelocity,
+    const ElementalData<NumNodes, Dim>& rData)
+{
+    // Compute density
+    const double local_mach_number_squared = PotentialFlowUtilities::ComputeLocalMachNumberSquared<Dim, NumNodes>(rVelocity, rCurrentProcessInfo);
+    const double density = PotentialFlowUtilities::ComputeDensity<Dim, NumNodes>(local_mach_number_squared, rCurrentProcessInfo);
+
+    // Compute density derivative
+    const double DrhoDu2 = PotentialFlowUtilities::ComputeDensityDerivativeWRTVelocitySquared<Dim, NumNodes>(local_mach_number_squared, rCurrentProcessInfo);
+
+    // Compute lhs
+    const BoundedVector<double, NumNodes> DNV = prod(rData.DN_DX, rVelocity);
+
+    rLhs_total = rData.vol * density * prod(rData.DN_DX, trans(rData.DN_DX));
+
+    const double local_velocity_squared = inner_prod(rVelocity, rVelocity);
+
+    const double max_velocity_squared = PotentialFlowUtilities::ComputeMaximumVelocitySquared<Dim, NumNodes>(rCurrentProcessInfo);
+    if (local_velocity_squared < max_velocity_squared){
+        rLhs_total += rData.vol * 2 * DrhoDu2 * outer_prod(DNV, trans(DNV));
```

Comment:
Yes, this is basically what we were missing for the clamped elements and it actually allows to have M >1

Reply:
Cool!
Btw, any reason to use the velocity squared over the mach number to check this condition?

I see this depends on the user define MACH_SQUARED_LIMIT. Wouldn''t it be better to have everything depending on a single parameter? (e.g MACH_LIMT only, and derive the rest from this)

Reply:
Good point, I think I see what you mean but I don't see the way of using the mach number to check this condition. The free stream velocity and mach number define the vacuum velocity, at which the local mach number becomes infinite and above which it is not defined. This is why we don't allow the mach number to go above certain value and the clamped elements do not "see" their real local mach number. This is the reason why I chose to compare the velocities instead of the mach numbers. We can discuss it in the next meeting, maybe I'm missing something.

Reply:
What I mean is that when computing the MaximumVelocitSquared, we are using:

https://github.com/KratosMultiphysics/Kratos/blob/77af460d6e1525e4cf07ddefec73eac8b693a77f/applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp#L158

Which is a user defined quantity. 

On one hand, since we also have the `MACH_LIMIT` as a user defined paramter, I wonder if we can remove the squared one, and do here ` MACH_LIMIT**2`. Maybe this is coming from literature?

On the other hand, my question was then: why not comparing the `local_mach_number < MACH_LIMIT`? Wouldn't it be the same? meaning that we dont allow local_mach numbers higher than this value.

Reply:
> On one hand, since we also have the `MACH_LIMIT` as a user defined paramter, I wonder if we can remove the squared one, and do here ` MACH_LIMIT**2`. Maybe this is coming from literature?

I agree we should only have one parameter. I would actually keep  ` MACH_LIMIT**2` because it is the term appearing in all of the formulas. This is the reason why in the transonic solver the limit is usually also set in the literature to  ` MACH_LIMIT**2 = 3.0` instead of using ` MACH_LIMIT`.  I didn't remove ` MACH_LIMIT` because it is still used in the non-perturbated compressible element. Do you have anything against keeping ` MACH_LIMIT**2` and removing  ` MACH_LIMIT`? Shall I adapt the non-perturbated compressible element to this?



> On the other hand, my question was then: why not comparing the `local_mach_number < MACH_LIMIT`? Wouldn't it be the same? meaning that we dont allow local_mach numbers higher than this value.

It is not the same. To make the comparison `local_mach_number < MACH_LIMIT` we would need to compute the  `local_mach_number` and this is not always possible, as we would get nan for high velocities. Let me try to explain what I mean with some plots. This is how the local mach number looks like vs the velocity magnitude:

![Screenshot from 2021-03-08 11-16-05](https://user-images.githubusercontent.com/28628414/110307868-b8474a80-7fff-11eb-954a-e98561c4250c.png)

As you can see, when approaching the velocity 807 m/s  the local mach number tends to infinity and beyond 807 m/s  it is not defined (807 m/s is the vacuum speed where the isentropic formulas that we are using are already way beyond their limits of application). In 2D such large velocities are usually not reached, but in 3D it can easily happen near the wing-tip trailing edge (it already happened to me when running some 3d examples).

This happens because the local mach number squared becomes negative beyond 807 m/s (which is of course physically impossible):

![Screenshot from 2021-03-08 11-24-03](https://user-images.githubusercontent.com/28628414/110308780-d2cdf380-8000-11eb-8e24-730e4fc1536f.png)

And this happens because the speed of sound squared becomes negative beyond the vacuum speed:

![Screenshot from 2021-03-08 11-27-52](https://user-images.githubusercontent.com/28628414/110309213-5b4c9400-8001-11eb-909f-6c539a147533.png)

This is by the way the reason why we need to introduce a limit on the mach number on the first place. The velocity magnitude squared, however, is always well defined:

![Screenshot from 2021-03-08 11-32-30](https://user-images.githubusercontent.com/28628414/110309794-0fe6b580-8002-11eb-8cfe-490fa60304dd.png)

Which is the reason why the velocity is one used for the checks instead of the mach number (including inside the computation of the local mach number itself):

https://github.com/KratosMultiphysics/Kratos/blob/293aab0852cdc8c642e010a1e5e5ca84b7f40555/applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp#L179-L188

Having said this, I think we need to improve some stuff:

- Remove one of the limits as you suggested.
- ComputeMaximumVelocitySquared only once and save it in the processinfo (Ill open another PR for this)
- What we do in [ComputeLocalMachNumber](https://github.com/KratosMultiphysics/Kratos/blob/77af460d6e1525e4cf07ddefec73eac8b693a77f/applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp#L502-L513) is very dangerous because it will give nan in 3D. So I would remove:  [ComputeLocalMachNumber](https://github.com/KratosMultiphysics/Kratos/blob/77af460d6e1525e4cf07ddefec73eac8b693a77f/applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp#L502-L513) and [ComputeLocalSpeedOfSound](https://github.com/KratosMultiphysics/Kratos/blob/77af460d6e1525e4cf07ddefec73eac8b693a77f/applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp#L404-L427) (I can also do this in another PR and adapt the non-perturbated elements to the safe way of computing the local mach number)
- Also at some point I would remove the non-perturbated elements and keep only the perturbated ones to clean up a bit (but I think we still have all of the embedded elements and the adjoints depending on them so we can do this later)

What do you think? Does this make sense to you? Maybe you have other comments? (I am pretty sure  we can improve and refactor the application a lot and I would like to do this once we have the 3D elements running)

Reply:
Sorry for the above long answer, I didn't know how to make it shorter :sweat_smile:

Reply:
Thanks for your detailed answer! It makes sense. 

> - Remove one of the limits as you suggested.

My only concern is a "practical" one, as it more intuitive to choose (as a user) a flat mach_number rather than a squared value. If in the code we keep the MACH_LIMIT**2 check, do you think it makes sense to make the user input the mach_numer limit (not squared). Or would you avoid this?

> - ComputeMaximumVelocitySquared only once and save it in the processinfo (Ill open another PR for this)
>- What we do in ComputeLocalMachNumber is very dangerous because it will give nan in 3D. So I would remove: ComputeLocalMachNumber and ComputeLocalSpeedOfSound (I can also do this in another PR and adapt the non-perturbated elements to the safe way of computing the local mach number)

Ok to me adressing this in a different PR

> Also at some point I would remove the non-perturbated elements and keep only the perturbated ones to clean up a bit (but I think we still have all of the embedded elements and the adjoints depending on them so we can do this later)

I agree with this but as you say the embedded/adjoint are depending on the un-perturbed elements and it will require some work to adapt and make sure eveyrthing is working correclty.

Reply:
> My only concern is a "practical" one, as it more intuitive to choose (as a user) a flat mach_number rather than a squared value. If in the code we keep the MACH_LIMIT**2 check, do you think it makes sense to make the user input the mach_numer limit (not squared). Or would you avoid this?

I know what you mean and honestly I don't know what is the best solution. From one point of view, I agree with you that thinking in terms of a flat mach_number is more intuitive and also in the purely subsonic compressible element the limit is usually set directly on the mach number. For example Eller proposes to use M=0.94, although for the Lock cases I was usually using M=0.99. For the transonic cases however since usually MACH_LIMIT**2=3.0 is used, I find this easier rather than having to input MACH_LIMIT=sqrt(3.0)=1.73205... Which for me also raises the question, which mach limit (or mach limit squared) to set as a default since we have two different values depending whether we are computing subsonic or transonic... Shall we move this setting to the [solver were we set up the element](https://github.com/KratosMultiphysics/Kratos/blob/77af460d6e1525e4cf07ddefec73eac8b693a77f/applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py#L46-L101)?

I would suggest that for now we remove MACH_LIMIT**2 from the input as you say and we move this setting to the set up of the element since it is element dependent, would you agree with this?

> Ok to me adressing this in a different PR

Ok, I'll try to do it this week

> I agree with this but as you say the embedded/adjoint are depending on the un-perturbed elements and it will require some work to adapt and make sure eveyrthing is working correclty.

Makes sense to me :+1: